src/Microsoft.IdentityModel.Tokens.Saml/SamlTokenUtilities.cs

```diff
@@ -102,5 +100,65 @@ internal static IEnumerable<SecurityKey> GetKeysForTokenSignatureValidation(stri
             }
             return null;
         }
+
+        /// <summary>
+        /// Creates <see cref="Claim"/>'s from <paramref name="claimsCollection"/>.
+        /// </summary>
+        /// <param name="claimsCollection"> A dictionary that represents a set of claims.</param>
+        /// <returns> A collection of <see cref="Claim"/>'s created from the <paramref name="claimsCollection"/>.</returns>
+        internal static IEnumerable<Claim> CreateClaimsFromDictionary(IDictionary<string, object> claimsCollection)
+        {
+            List<Claim> claims = null;
+            if (claimsCollection == null)
+                return claims;
+
+            claims = new List<Claim>();
+            string valueType;
+            foreach (var claim in claimsCollection)
+            {
+                string claimType = claim.Key;
+                object claimValue = claim.Value;
+
+                if (claimValue != null)
+                {
+                    valueType = TokenUtilities.GetClaimValueTypeFromValue(claimValue);
+                    if (valueType == null)
+                    {
+                        if (claimValue is IEnumerable claimList)
+                        {
+                            foreach (var item in claimList)
+                            {
+                                valueType = TokenUtilities.GetClaimValueTypeFromValue(item);
+                                claims.Add(new Claim(claimType, item.ToString(), valueType == null? item.GetType().ToString() : valueType));
```

Comment:
@RojaEnnam 

the valueType must be a valid XSI type { "http://www.w3.org/2001/XMLSchema#string" is assumed}
see: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/8ca7558ca2d5c7a27d108d6ecd7c9babf2b30a95/src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2Serializer.cs#L1639 and https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/8ca7558ca2d5c7a27d108d6ecd7c9babf2b30a95/src/Microsoft.IdentityModel.Tokens.Saml/Saml/SamlAttribute.cs#L92

If we add the following to CreateSaml2TokenUsingTokenDescriptor and CreateSamlTokenUsingTokenDescriptor tests, they will throw on creation.

{ "https://wwww.listinlist.com", new List<object>{ new List<object> { "bob", new SecurityTokenDescriptor(), 12, 1.45 }, new List<object> { "bob", new SecurityTokenDescriptor(), 12, 1.45 } } }

We should throw a NotSupportedException for enumerations inside enumerations and leave the valueType to null if we can not resolve it to a well known xsiType from GetClaimValueType. 

Since this type is only for Saml tokens, how about renaming GetClaimValueTypeFromValue to GetXsiTypeForValue and move the method to SamlTokenUtilities.

Reply:
I think that makes it more clear. Pushed the changes.